tracy-client/src/gpu.rs

```diff
@@ -0,0 +1,201 @@
+    context: u8,
+    #[cfg(feature = "enable")]
+    // This is the index first query, the second query is always a wrapping_add(1) above this.
+    query_id: u16,
+    #[cfg(feature = "enable")]
+    ended: bool,
+}
+
+impl Client {
+    /// Creates a new GPU context.
+    ///
+    /// - 'name' is the name of the context.
+    /// - 'ty' is the type (backend) of the context.
+    /// - 'gpu_timestamp' is the gpu side timestamp the corresponds (as close as possible) to this call.
+    /// - 'period' is the period of the gpu clock in nanoseconds (setting 1.0 means the clock is 1GHz, 1000.0 means 1MHz, etc).
+    pub fn new_gpu_context(
+        self,
+        name: Option<&str>,
+        ty: GpuContextType,
+        gpu_timestamp: i64,
+        period: f32,
+    ) -> Option<GpuContext> {
+        #[cfg(feature = "enable")]
+        {
+            // We use a mutex to lock the context index to prevent races when using fetch_add.
+            let mut context_index_guard = GPU_CONTEXT_INDEX.lock().unwrap();
+            if *context_index_guard == 255 {
+                return None;
+            }
+            let context = *context_index_guard;
+            *context_index_guard += 1;
+            drop(context_index_guard);
+
+            unsafe {
+                sys::___tracy_emit_gpu_new_context(sys::___tracy_gpu_new_context_data {
+                    gpuTime: gpu_timestamp,
+                    period,
+                    context,
+                    flags: 0,
+                    type_: ty as u8,
+                })
+            };
+
+            if let Some(name) = name {
+                unsafe {
+                    sys::___tracy_emit_gpu_context_name_serial(
+                        sys::___tracy_gpu_context_name_data {
+                            context,
+                            name: name.as_ptr().cast(),
+                            len: name.len().try_into().unwrap_or(u16::MAX),
+                        },
+                    )
+                }
+            }
+
+            Some(GpuContext {
+                client: self,
+                context,
+                span_index: Arc::new(AtomicU16::new(0)),
+            })
+        }
+        #[cfg(not(feature = "enable"))]
+        Some(GpuContext {})
+    }
+}
+
+impl GpuContext {
+    ///
+    pub fn span_alloc(&self, name: &str, function: &str, file: &str, line: u32) -> GpuSpan {
+        #[cfg(feature = "enable")]
+        {
+            let srcloc = unsafe {
+                sys::___tracy_alloc_srcloc_name(
+                    line,
+                    file.as_ptr().cast(),
+                    file.len(),
+                    function.as_ptr().cast(),
+                    function.len(),
+                    name.as_ptr().cast(),
+                    name.len(),
+                )
+            };
+
+            // Allocate two ids, one for the start value, one for the end.
+            let query_id = self.span_index.fetch_add(2, Ordering::Relaxed);
```

Comment:
When are these IDs freed?

Reply:
My understanding is once the spans ids have timestamps associated with them they can be reused again. The consequences of reusing queries are not exactly clear to me (this api has basically no documentation), and we might need to add a freelist to ensure the is no improper reuse.

Reply:
At a guess, perhaps `___tracy_emit_gpu_time_serial` is an implicit dealloc? In that case we'd need to do something in `Drop`.

Reply:
I'm not sure what we'd do in a drop. Maybe write out junk zero-value timestamps? Or write out the same timestamp as the start of the gpu context, so they don't interfere with other spans?

Reply:
The second sounds like the safest bet to me, seeing as Tracy's C++ API tries to make this case impossible.

Reply:
Done.